### PR TITLE
Add FlyC:0 and updated the mode menu

### PR DESCRIPTION
**Pull request by JohnC32**
1. Add FlyC:0

   The existing 'FlyC' mode line indicator had two meanings
   - current buffer has not been checked, or
   - last check resulted in no warnings and no errors and the manual indicated that it was the later which I found confusing.

   This commit, defines:

   FlyC     This buffer has not been checked yet.
   FlyC:0   Last check resulted and no errors and no warnings.

   Also add colors to the mode line status indicator:

   - FlyC:0     use 'success face
   - FlyC:3|5   use 'error face
   - FlyC!      use 'error face

2. Update the mode menu to add

   - "Run checker as compile command" which is `flycheck-compile'. This is helpful for multiline errors coming from compilers such as clang. Also updated the `flycheck-compile' prompt to show the default checker when there is one because typing only RET will run the default checker.

   - "Flycheck quick help" which runs new sub, `flycheck-quick-help' and displays what the mode line indicators mean along with some of the more common key bindings.

### CODE REVIEW

**Comment by JohnC32** [comment]
> [CLA assistant check](https://cla-assistant.io/flycheck/flycheck?pullRequest=2035) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/flycheck/flycheck?pullRequest=2035) before we can accept your contribution.
> 
> **John Ciolfi** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).
> You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/flycheck/flycheck?pullRequest=2035) it.

I signed it and my account is associated with an e-mail. Not sure what's up. Let me know if you want the fixes, if not I'll delete the PR.

**Comment by bbatsov** [comment]
@JohnC32 Sorry for the slow response. Your PR seems reasonable to me and I'll review it in more detail soon.

**Comment by bbatsov** [comment]
> Also add colors to the mode line status indicator:

I'd probably make this an optional feature controlled by a `defcustom`. 

Overall your changes look pretty good to me. Please, mention them in our changelog. It would also be nice if you add a couple of buttercup tests that cover them.

**Comment by bbatsov** [comment]
@JohnC32 Friendly ping :-)

**Comment by JohnC32** [comment]
Sorry for the delay, I just noticed the ping. We are using flycheck heavily and really like it. Thanks for maintaining it.

I added defcustom for the flycheck mode line status coloring. 

I saw the comet about wanting two separate PR's. I'm not sure how to split this into two PRs without deleting my fork and then re-doing it. If it's okay, can you take it as is?

Thanks
John

File: flycheck.el [hunk 1]
```diff
@@ -2936,10 +2980,11 @@ Flycheck displays its status in the mode line.  In the default
 configuration, it looks like this:
 
 `FlyC'     This buffer has not been checked yet.
-`FlyC-'    Flycheck doesn't have a checker for this buffer.
 `FlyC*'    Flycheck is running.  Expect results soon!
-`FlyC:3|2' This buffer contains three warnings and two errors.
+`FlyC:0'   Last check resulted and no errors and no warnings.
```
Comment (bbatsov) [inline flycheck.el:2984]: resulted and -> resulted in

File: flycheck.el [hunk 2]
```diff
@@ -1284,6 +1292,42 @@ just return nil."
...
+  (with-current-buffer (get-buffer-create "*flycheck-quick-help*")
+    (with-help-window (current-buffer)
+      (flycheck-mode) ;; so that we can exapnd \\[flycheck-<function>]
+      (let ((help
+             (substitute-command-keys
+        "Flycheck automatically runs checks on writable files when changed.
+Mode line status for the current buffer:
+  FlyC        Not been checked yet
+  FlyC*       Flycheck is running
+  FlyC:0      Last check resulted and no errors and no warnings
```
Comment (bbatsov) [inline flycheck.el:1308]: resulted in

Comment (JohnC32) [reply flycheck.el]: fixed typo

File: flycheck.el [hunk 3]
```diff
@@ -4095,19 +4140,29 @@ refresh the mode line."
...
-                   (if (or .error .warning)
-                       (format ":%s|%s" (or .error 0) (or .warning 0))
-                     "")))
-                (`interrupted ".")
-                (`suspicious "?"))))
-    (concat " " flycheck-mode-line-prefix text)))
+  (let (text
+        face)
+    (setq text (concat " " flycheck-mode-line-prefix
+                       (pcase (or status flycheck-last-status-change)
```
Comment (bbatsov) [inline flycheck.el:4146]: Can you move the `pcase` out of the `concat`, as I think the way it's current written the readability is not optimal? (plus it shift the indentation too much to the right)

File: flycheck.el [hunk 4]
```diff
@@ -4095,19 +4140,29 @@ refresh the mode line."
...
+                          (let-alist (flycheck-count-errors flycheck-current-errors)
+                            (if (or .error .warning)
+                                (progn
+                                  (setq face 'error)
+                                  (format ":%s|%s" (or .error 0) (or .warning 0)))
+                              (setq face 'success)
+                              ":0")))
+                         (`interrupted ".")
+                         (`suspicious "?"))))
+    (when (and face flycheck-mode-line-color)
```
Comment (bbatsov) [inline flycheck.el:4163]: I'm guessing you can just use `if/else` here and have the normal and colorized version of the text as the two branches. (or something along those lines) I'd also suggest using `format` over `concat`.

**Comment by bbatsov** [comment]
> I saw the comet about wanting two separate PR's. I'm not sure how to split this into two PRs without deleting my fork and then re-doing it. If it's okay, can you take it as is?

Yeah, that'd be fine. I felt a bit more small remarks today while reviewing the code more carefully.

You'll also need to mention the new config option somewhere in the user docs.

File: flycheck.el [hunk 5]
```diff
@@ -4095,19 +4140,29 @@ refresh the mode line."
...
+                                     (setq face 'error)
+                                     "!"))
+                         (`finished
+                          (let-alist (flycheck-count-errors flycheck-current-errors)
+                            (if (or .error .warning)
+                                (progn
+                                  (setq face 'error)
+                                  (format ":%s|%s" (or .error 0) (or .warning 0)))
+                              (setq face 'success)
+                              ":0")))
```
Comment (bbatsov) [inline flycheck.el:4160]: Btw, why don't we make this something configurable so people can tweak it if they want to? E.g. `flycheck-mode-success-indicator` (a string). I think that'd be best as some people like like `0`, some `0:0`, and some might want something like a checkmark symbol or an "OK".

**Comment by bbatsov** [comment]
@JohnC32 friendly ping :-)

File: flycheck.el (hunk 4 above)
Comment (JohnC32) [reply flycheck.el]: Separated the mode indicator text from the face selection

File: flycheck.el (hunk 5 above)
Comment (JohnC32) [reply flycheck.el]: Added `flycheck-mode-success-indicator'

**Comment by bbatsov** [comment]
The changes look good overall, but it seems they broke a couple of the unit tests. Those will need to be updated.

**Comment by bbatsov** [comment]
```
Mode Line does not show the number of infos

Traceback (most recent call last):
  buttercup-fail("%s" "Expected `(flycheck-mode-line-status-text (quote fini...
  signal(buttercup-failed "Expected `(flycheck-mode-line-status-text (quote ...
FAILED: Expected `(flycheck-mode-line-status-text (quote finished))' to be `equal' to `" FlyC"', but instead it was `#(" FlyC:0" 0 7 (face success))' which does not match because: (arrays-of-different-length 7 5 #(" FlyC:0" 0 7 (face success)) " FlyC" first-mismatch-at 5).

========================================
Mode Line includes the prefix

Traceback (most recent call last):
  buttercup-fail("%s" "Expected `(flycheck-mode-line-status-text (quote fini...
  signal(buttercup-failed "Expected `(flycheck-mode-line-status-text (quote ...
FAILED: Expected `(flycheck-mode-line-status-text (quote finished))' to be `equal' to `" foobar"', but instead it was `#(" foobar:0" 0 9 (face success))' which does not match because: (arrays-of-different-length 9 7 #(" foobar:0" 0 9 (face success)) " foobar" first-mismatch-at 7).

Ran 113 specs, 2 failed, in 2.49s.
```

**Comment by bbatsov** [comment]
@JohnC32 ping :-)

**Comment by JohnC32** [comment]
> @JohnC32 ping :-)

Sorry for the delay - been too busy with other things. I did try to get the tests running on a Debian 11 system, but was unable to do so? Looking at the failures, the fix seemed very straightforward.

**Comment by bbatsov** [comment]
Thanks for driving this to the finish line!

**Pull request merged**